api/v1alpha1/etcdpeer_types.go

```diff
@@ -7,10 +7,26 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// BootstrapPeer describes a single *other* peer that will be used for bootstrapping
+type BootstrapPeer struct {
+	// Name is the initial name of the peer, which is set into etcd at the
+	// end of this process
+	Name string `json:"name"`
+
+	// Host forms part of the Advertise URL - the URL at which this peer can
+	// be contacted. The port and scheme are hardcoded.
```

Comment:
"the port and scheme are hardcoded" I don't think they should be. Let's have the `host` here be e.g. `https://example.com:2380`. The EtcdPeer should be a fairly agnostic resource, and should not dictate things like schemes and port numbers. It's a wrapper around the Pod API mostly, with some extra spec/status fields 😄

Reply:
I did think about that, but I was worried about how to do things like default the scheme or port if it's not provided.

> The EtcdPeer should be a fairly agnostic resource, and should not dictate things like schemes and port numbers. It's a wrapper around the Pod API mostly, with some extra spec/status fields

Isn't the point of the `EtcdPeer` resource encapsulate the management and configuration of a single instance of etcd away from the cluster controller? Exactly so that the cluster controller does not concern itself with things like port numbers, schemes, or how to provide information to the etcd container. So quite a bit more than just a wrapper around `Pod`.

Reply:
In terms of the hard coding. I was intending for the `BootstrapPeer` struct to, in future, include a `Port` and `scheme` optional field.

Reply:
> Isn't the point of the EtcdPeer resource encapsulate the management and configuration of a single instance of etcd away from the cluster controller?

Right, but I don't think this really helps here. We are referencing *other members* in a cluster, and by hard coding/deriving the scheme/port number of the EtcdPeer controller, the *peer* is now making an assumption about the *clusters* configuration, which is definitely incorrect.

> I was intending for the BootstrapPeer struct to, in future, include a Port and scheme optional field.

It sounds like you agree then that it *shouldn't* be hardcoded longer-term. I am not convinced that making these standalone fields is actually more beneficial than having strings containing FQDN. Parsing URL schemes and ports, if required, is a trivial and well understood operation. I don't think it's particular beneficial to default at the EtcdPeer level in this way. I think generic URLs are a closer representation of what etcd itself actually accepts as configuration.

How about we go with using the full URLs for now (e.g. `https://example.com:2380`), passing these strings straight through to etcd's env/args. If we find a need for explicit Scheme and Port fields in future, we can add them in (with corresponding validation to ensure that `port` as a field isn't specified along with an explicit port in the URL).

Reply:
> Parsing URL schemes and ports, if required, is a trivial and well understood operation. I don't think it's particular beneficial to default at the EtcdPeer level in this way. I think generic URLs are a closer representation of what etcd itself actually accepts as configuration.

I can see it. I'll change to my initial implementation of having a `Name` and a `AdvertiseURL` on the `BootstrapPeer` struct.

Reply:
i find explicit port/scheme/address to be helpful as it makes it clear what is required. it's easy to assume that `Host`  may only need to be an IP, for example, and may lead to people getting trolled. also makes things a bit more consistent with the rest of the k8s api (probes)

Reply:
Thinking about it, I think you wouldn't ever have a path so it's not really an arbitary FQDN. I'm unsure that you can specify any scheme other than http either.

I should probably use the Go URL library to assemble it too, rather than `fmt.Sprintf`.

Reply:
> I'm unsure that you can specify any scheme other than http either.

HTTPS? 🙃 

> I should probably use the Go URL library to assemble it too, rather than fmt.Sprintf.

:+1:

Reply:
I've also realised that you don't need to specify yourself in this list. The peer operator already needs to know the peer's own advertisement URL, so it can add it itself.

Reply:
I thought I mentioned this before, but I'm personally not convinced omitting it is something we should allow. Notably because...

1) If a user does *not* specify it, it will be automatically 'injected'
2) If a user *does* specify it however, what do we then do? Deduplicate? This makes the logic inside the EtcdPeer controller more complex, as it then has to discover if one already exists. What if just the schemes differ? What if just one part of it differs?

You are effectively taking control away from the user and making the API slightly less flexible, and introducing a cost of having to compute unique member lists for each peer.

I think we can change this in future either way, so I'm not *too* fussed, but personally I'm strongly in favour of *requiring* that the full member list is provided. Adding a new peer becomes non-trivial and complex, and checking to make sure a peer *does* have the correct members configured is also less obvious too (every member list on every peer will be different). Given that ultimately etcd requires *all* specified, I just don't see much value here 🤷‍♂

Reply:
Given we sort of *have* to allow users *to* specify it, how about to start with we do require it be specified, and in future if this becomes a pain point, we can relax this validation to allow it to be omitted?

Reply:
I've made it so you need to give it.

Reply:
So I've made it that you have to specify all members. Are we happy with just `name` and `host` for now, with the expectation that we will add `port` and `scheme` as additional optional fields at a future date?

Reply:
Sure - I think we'll need those additional fields fairly soon, but SGTM 😄